File: server/src/main/java/gg/mineral/server/util/nbt/NBTReadLimiter.java
package gg.mineral.server.util.nbt;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class NBTReadLimiter {

    public static final NBTReadLimiter UNLIMITED = new NBTReadLimiter(0L) {

        @Override
        public void read(int length) {
        }
    };

    private final long limit;
    private long currentRead = 0;

    public void read(int length) {
        this.currentRead += length;
        if (this.currentRead > this.limit)
            throw new IllegalStateException("Read more than " + this.limit + " bytes from NBT tag");
    }
}

File: api/src/main/kotlin/gg/mineral/api/nbt/NBTReadLimiter.kt
package gg.mineral.api.nbt

open class NBTReadLimiter(val limit: Long) {
    private var currentRead: Long = 0

    open fun read(length: Int) {
        this.currentRead += length.toLong()
        check(this.currentRead <= this.limit) { "Read more than " + this.limit + " bytes from NBT tag" }
    }

    companion object {
        @JvmField
        val UNLIMITED: NBTReadLimiter = object : NBTReadLimiter(0L) {
            override fun read(length: Int) {}
        }
    }
}
